fixes logic in locale redirect

Adebesin-Cell opened the pull request:
# fixes logic in locale redirect

This PRs handles edge cases causes multiple redirections in site

Fixes https://github.com/EveripediaNetwork/issues/issues/2172

Royal-lobster commented:
@Adebesin-Cell it still doing infinite redirect, try this out

```ts
export function middleware(req: NextRequest) {
  const isMaintenanceMode = process.env.MAINTENANCE_MODE === 'true';

  // Check for maintenance mode first
  if (
    isMaintenanceMode &&
    req.nextUrl.pathname !== '/maintenance' &&
    !req.nextUrl.pathname.includes('.')
  ) {
    const url = req.nextUrl.clone();
    url.pathname = '/maintenance';
    return NextResponse.redirect(url.toString(), { status: 302 });
  }

  const userLocale = req.headers
    .get('accept-language')?
    .split(',')[0]
    .split('-')[0];

  // Check if the requested pathname already has a supported locale
  const isLocaleAlreadySet = languageData.some((lang) =>
    req.nextUrl.pathname.startsWith(`/${lang.locale}/`)
  );

  // If the user's locale is already set in the pathname, do not redirect
  if (isLocaleAlreadySet) {
    return NextResponse.next();
  }

  // Redirect based on user's locale if not already in the pathname
  const isLocaleSupported = languageData.some(
    (lang) => lang.locale === userLocale
  );

  // Redirect to user's locale if supported and not already in the pathname
  if (!isLocaleAlreadySet && isLocaleSupported) {
    const url = req.nextUrl.clone();
    url.pathname = `/${userLocale}${req.nextUrl.pathname}`;
    return NextResponse.redirect(url.toString(), { status: 302 });
  }

  // Redirect to default locale if user's locale is not supported and not already in the pathname
  if (!isLocaleAlreadySet && !isLocaleSupported) {
    const defaultLocale =
      languageData.find((lang) => lang.default)?.locale ?? 'en';
    const url = req.nextUrl.clone();
    url.pathname = `/${defaultLocale}${req.nextUrl.pathname}`;
    return NextResponse.redirect(url.toString(), { status: 302 });
  }

  return NextResponse.next();
}
```

Adebesin-Cell commented:
> @Adebesin-Cell it still doing infinite redirect, try this out
> 
> ```ts
> export function middleware(req: NextRequest) {
>   const isMaintenanceMode = process.env.MAINTENANCE_MODE === 'true';
> 
>   // Check for maintenance mode first
>   if (
>     isMaintenanceMode &&
>     req.nextUrl.pathname !== '/maintenance' &&
>     !req.nextUrl.pathname.includes('.')
>   ) {
>     const url = req.nextUrl.clone();
>     url.pathname = '/maintenance';
>     return NextResponse.redirect(url.toString(), { status: 302 });
>   }
> 
>   const userLocale = req.headers
>     .get('accept-language')?
>     .split(',')[0]
>     .split('-')[0];
> 
>   // Check if the requested pathname already has a supported locale
>   const isLocaleAlreadySet = languageData.some((lang) =>
>     req.nextUrl.pathname.startsWith(`/${lang.locale}/`)
>   );
> 
>   // If the user's locale is already set in the pathname, do not redirect
>   if (isLocaleAlreadySet) {
>     return NextResponse.next();
>   }
> 
>   // Redirect based on user's locale if not already in the pathname
>   const isLocaleSupported = languageData.some(
>     (lang) => lang.locale === userLocale
>   );
> 
>   // Redirect to user's locale if supported and not already in the pathname
>   if (!isLocaleAlreadySet && isLocaleSupported) {
>     const url = req.nextUrl.clone();
>     url.pathname = `/${userLocale}${req.nextUrl.pathname}`;
>     return NextResponse.redirect(url.toString(), { status: 302 });
>   }
> 
>   // Redirect to default locale if user's locale is not supported and not already in the pathname
>   if (!isLocaleAlreadySet && !isLocaleSupported) {
>     const defaultLocale =
>       languageData.find((lang) => lang.default)?.locale ?? 'en';
>     const url = req.nextUrl.clone();
>     url.pathname = `/${defaultLocale}${req.nextUrl.pathname}`;
>     return NextResponse.redirect(url.toString(), { status: 302 });
>   }
> 
>   return NextResponse.next();
> }
> ```

How did you replicate the multiple redirect?

Royal-lobster commented:
You can change your browser language, search for korean and make it to top

![CleanShot 2024-01-05 at 15 00 33@2x](https://github.com/EveripediaNetwork/ep-ui/assets/52039218/2ffff2b7-8704-45b2-a939-35969f421ac3)

Adebesin-Cell commented:
It keeps redirecting, I removed the logic and it seems to work fine aside the fact that the url doesn't change if the locale isn't present for translation.

Royal-lobster commented:
> It keeps redirecting, I removed the logic and it seems to work fine aside the fact that the url doesn't change if the locale isn't present for translation.

yeah, i think nextjs already has the feature to detect the language of user right ? https://nextjs.org/docs/pages/building-your-application/routing/internationalization i believe we don't need to touch middleware

Adebesin-Cell commented:
> > It keeps redirecting, I removed the logic and it seems to work fine aside the fact that the url doesn't change if the locale isn't present for translation.
> 
> yeah, i think nextjs already has the feature to detect the language of user right ? https://nextjs.org/docs/pages/building-your-application/routing/internationalization i believe we don't need to touch middleware

Yep! they handle it.

Royal-lobster commented:
> Yep! they handle it.

Cool, make the change and feel free to ping here back so i can help test

Adebesin-Cell commented:
> > Yep! they handle it.
> 
> Cool, make the change and feel free to ping here back so i can help test

https://nextjs.org/docs/pages/building-your-application/routing/internationalization#leveraging-the-next_locale-cookie

I'm using this rn in the system during language switch. And the system works by giving language locale it detects it renders based on that. You can test and see if that works or needs improvement.

Royal-lobster commented:
@Adebesin-Cell Now its not detecting my browser language, its showing english version with prefered korean set in settings

Royal-lobster commented on next-i18next.config.js:
```diff
@@ -4,7 +4,7 @@ const path = require('path')
   i18n: {
     defaultLocale: 'en',
-    locales: ['en', 'kr', 'zh'],
+    locales: ['en', 'kr'],
```
I believe we should also add this to next.config.js

![CleanShot 2024-01-08 at 14 03 30@2x](https://github.com/EveripediaNetwork/ep-ui/assets/52039218/81cd9830-5015-4c67-9782-619bd8bcde75)

Adebesin-Cell replied:
I think it was abstracted into a new file and imported.

Adebesin-Cell commented:
> @Adebesin-Cell Now its not detecting my browser language, its showing english version with prefered korean set in settings

I'll look into that.

Royal-lobster commented:
Its wierd that [automatic locale detection](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection) is not working :/ 

But check out this code, I made it work with middleware
https://github.com/Royal-lobster/Nextjs-Locale-Routing

![CleanShot 2024-01-09 at 12 53 36](https://github.com/EveripediaNetwork/ep-ui/assets/52039218/16053e14-21d1-4964-923c-6eaf7669136f)

I think you are missing this step, when you use middleware, not only your current page gets passed to it, but also assets such as favicons and stuff. if you don't stop redirecting them, it goes into loop

![CleanShot 2024-01-09 at 12 53 17](https://github.com/EveripediaNetwork/ep-ui/assets/52039218/ddf284a5-8453-4908-8a0a-ab6d39a9a561)

Adebesin-Cell commented:
> Its wierd that [automatic locale detection](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection) is not working :/
> 
> But check out this code, I made it work with middleware https://github.com/Royal-lobster/Nextjs-Locale-Routing
> 
> ![CleanShot 2024-01-09 at 12 53 36](https://private-user-images.githubusercontent.com/52039218/295130586-16053e14-21d1-4964-923c-6eaf7669136f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MTEzMzUsIm5iZiI6MTcwNDgxMTAzNSwicGF0aCI6Ii81MjAzOTIxOC8yOTUxMzA1ODYtMTYwNTNlMTQtMjFkMS00OTY0LTkyM2MtNmVhZjc2NjkxMzZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDE0MzcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjYjcwOWM1NzEyMDgyZWE3YzVkOWUyYWYxM2MzNmUzNjI2YjUyNjY5YWQxMjViZDhiZTU0ZjM3ZWQ2YzMxNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6mPTLvXOup0zIBaXX2YVnK206ZT2BtQOzuEsGJcIyFY)
> 
> I think you are missing this step, when you use middleware, not only your current page gets passed to it, but also assets such as favicons and stuff. if you don't stop redirecting them, it goes into loop
> 
> ![CleanShot 2024-01-09 at 12 53 17](https://private-user-images.githubusercontent.com/52039218/295130648-ddf284a5-8453-4908-8a0a-ab6d39a9a561.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MTEzMzUsIm5iZiI6MTcwNDgxMTAzNSwicGF0aCI6Ii81MjAzOTIxOC8yOTUxMzA2NDgtZGRmMjg0YTUtODQ1My00OTA4LThhMGEtYWI2ZDM5YTlhNTYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDE0MzcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlM2UyNjc2ODdmY2ViNmJlNGUwN2I1MjNjMzBhMzY3MWI0ZTFjZTkzN2I2ZDg2YjA2NTg5MzkxMzBjZTI1ZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ESfDcbBEbbLFI8f1Mns1Ia3bmtWjHIRZgL2m6uLkbvE)

I tried this similar approach, it keeps giving "TOO MANY REDIRECTS"

Royal-lobster commented:
There is also another case, if user's preference is default we don't redirect cuz default is just / and it redirects in loop. pushed fix in the commit. 

![CleanShot 2024-01-10 at 10 46 40](https://github.com/EveripediaNetwork/ep-ui/assets/52039218/eae3014d-7af4-4040-b8b8-4f47e8805879)

Another thing that you need to do in this pr is, change kr to ko. AcceptedLanguages header passes korean preference as ko. but we are using kr as symbol for korean. currently its redirecting to 404

Adebesin-Cell commented:
> There is also another case, if user's preference is default we don't redirect cuz default is just / and it redirects in loop. pushed fix in the commit.
> 
> ![CleanShot 2024-01-10 at 10 46 40](https://private-user-images.githubusercontent.com/52039218/295456083-eae3014d-7af4-4040-b8b8-4f47e8805879.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4NjQyODQsIm5iZiI6MTcwNDg2Mzk4NCwicGF0aCI6Ii81MjAzOTIxOC8yOTU0NTYwODMtZWFlMzAxNGQtN2FmNC00MDQwLWI4YjgtNGY0N2U4ODA1ODc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDA1MTk0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViMjhlNjA3YWQ5NmI2Mjg4Y2E5OTE3YzUyOTI5YzliMWIxMDlmMWM4NDg0YjRiZDkzMjAwNjQ0NmY2ODk5N2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.K6U_pBuyPeKraY5_n2UKQcQABcscxDlP7G-7Tn758aU)
> 
> Another thing that you need to do in this pr is, change kr to ko. AcceptedLanguages header passes korean preference as ko. but we are using kr as symbol for korean. currently its redirecting to 404

yeah, there's a function for that. ``revertToKr``

Royal-lobster commented:
> yeah, there's a function for that. `revertToKr`

Oh. Didn't study that code, but i removed them from utils folder cuz we are not using them in middleware now

Adebesin-Cell commented:
Prolly related to this merged, we have cloudfare issues on [https://iq.wiki/activity](https://iq.wiki/ko/acitivity).
